### DISCUSSION TITLE
How to work with ParsedUrlQuery?

### DISCUSSION THREAD

**Discussion by florianwalther-private** [vercel/next.js#38061] [Help] [answered]
I'm converting from CRA, where I got my URL query params as a `URLSearchParams` object, to NextJS. In `getServerSideProps`, I can receive the query params as a `ParsedUrlQuery` object. Is there a simple way to convert this `ParsedUrlQuery` into `URLSearchParams` or at least url-encode them so I can send them as a query to the server (in the same form as I received them)?

**Comment by ramirezsandin** [comment]
Embrace the ParsedUrlQuery, here you have a hook to manage the query params in an easy way
https://gist.github.com/ramirezsandin/cc93f6a405ff818e48038e28a26a1ba0

**Reply by florianwalther-private** [reply]
Thank you. Although replacing 1 line for 230 lines is kind of a bummer.

**Comment by icyJoseph** [comment]
You can do:

```tsx
const Home: NextPage = (props) => {
  console.log(props);
  return (
    <div>
      Home <Link href="/blog">blog</Link>
    </div>
  );
};

export const getServerSideProps = (ctx: GetServerSidePropsContext) => {
  const { query } = ctx;

  const search = new URLSearchParams();

  Object.entries(query).forEach(([key, value]) => {
    search.append(key, `${value}`); // maybe you want to do `Array.isArray(value)` checks, etc
    // basically, serialize it as you wish based on its type
  });

  return { props: { search: search.toString() } }; // this bit is just to show them client side, I think you want to forward them elsewhere?
};

export default Home;
```

Then you could go to: `/?foo&bar=joe&hello=world&bar=jane`, and see:

```js
{search: 'foo=&bar=joe%2Cjane&hello=world'} // save this log a global temp1 in devTools
const search = new URLSearchParams(temp1.search);
search.get('bar') // prints 'joe,jane'
```

**Reply by florianwalther-private** [reply]
I think this will not work if an entry has more than one value

**Reply by icyJoseph** [reply]
My example shows two entries for `bar` :) but also `/?foo&bar=joe&hello=world,all&bar=jane` should work, with multiple bar entries and comma separated values for hello. Even `/?foo&bar=joe&hello=world,all&bar=jane,name` should work.

**Reply by florianwalther-private** [reply]
@icyJoseph Ah yea now I see it 👍 Thank you, I will try it out!

**Reply by ramirezsandin** [reply]
The querystring package is something that is considered legacy on Node, and they guys from Vercel should consider replace it with URLSearchParams

**Reply by florianwalther-private** [reply]
@icyJoseph Is there a specific reason you wrote  `` `${value}` `` instead of just `value`?

**Reply by florianwalther-private** [reply]
This does not work for me with multiple values for a single key. It just tries to url-encode the comma.

**Reply by ramirezsandin** [reply]
Here you have it, a direct transformation of types:
```typescript
export function parsedUrlQueryToURLSearchParams(
  query: ParsedUrlQuery
): URLSearchParams {
  const searchParams = new URLSearchParams()
  for (const [key, value] of Object.entries(query)) {
    if (!value) continue
    if (Array.isArray(value)) {
      value.forEach((element) => {
        searchParams.append(key, element)
      })
    } else {
      searchParams.append(key, value)
    }
  }
  return searchParams
}
export function urlSearchParamsToParsedUrlQuery(
  searchParams: URLSearchParams
): ParsedUrlQuery {
  const query: ParsedUrlQuery = {}
  for (var [key, value] of searchParams.entries()) {
    query[key] = value
  }
  return query
}
```
with URLSearchParams, if you do `searchParams.get('key')` you get the first value, if you do `searchParams.getAll('key')` and array with the values.

**Reply by florianwalther-private** [reply]
@ramirezsandin Thank you for your efforts! It's great. However, the fact that I need this tells me that I must be doing something wrong elsewhere. Sending query params to a server is a common use case. What am I missing? How are others doing it?

**Reply by icyJoseph** [reply]
I did `${value}` to cast the array to a list separated by commas...

I think we've gotten very side tracked here, I don't really understand where have we gone dark, but we are not communicating at all here.

When the user makes a request for your page that implements `getServerSideProps`, Next.js will make an object with the queries, the signature of the object is `Record<string, string | string[]>`, where and when reading, you must account for the fact that the query might not have been there to begin with, which means that:

```ts
const {query} = ctx;
const foo = query.foo; // could be string, string list, or undefined
```

`foo` can be string, a list of strings, or `undefined`.

- `string`: when there's only one foo in the browser url query
- string list: when there's multiple foo, one entry per `foo=` instance in the query
- `undefined`: when it was not present to begin with

Now, where do you want to go from there.

**Reply by florianwalther-private** [reply]
@icyJoseph I want to send the query params to the server. @ramirezsandin's code works. But it's odd that I need to create my own helper functions for such a common task. I must be missing something.

**Reply by icyJoseph** [reply]
> But it's odd that I need to create my own helper functions for such a common task I must be missing something.

How's that? What's the server in this case? Is it `getServerSideProps`? You don't need any functions to share URL queries with the server, as you say, it is an out of the box kind of feature.

**Reply by florianwalther-private** [reply]
@icyJoseph The server is an Express server and the query params are sent via the URL (The same way I received them). But there seems to be no simple way to url-encode `ParsedUrlQuery`. `URLSearchParams`, on the other hand, are url-encoded by just calling `toString()`.

**Reply by icyJoseph** [reply]
Alright, so what we want to do is forward the queries, as the browser sees them to another entity. So you can do:

```ts
 const url = new URL(ctx.resolvedUrl, "https://server-api.com"); // use the URL of your express server as second argument
 console.log(url.search);
 const search = url.search; // use this to make a fetch request against your server
```

There's a bunch of cool one lines you can pull of here depending on the api your search uses, but:

```ts
await fetch('https://server-api.com/resource/path' + url.search);
```

Is the most straight forward with the least amount of assumptions.

**Reply by ramirezsandin** [reply]
True is that sending query params to the server is a common case, but I think that's left open to the developer to choose how to do it. 

You were using the URLSearchParams, that's just a collection of utility methods to work with the strings on the URL, Next.js uses querystring and it's also valid (but doesn't have more helper methods).  You just have to adapt and use what is provided to you, probably there will be more libraries to work with that.

But basically for what I've seen in your api call function, you pass a variable named searchParams, of type URLSearchParams to the function right?, this variable has a method toString() that converts the object to a string and if I'm not wrong that is what it's read from the server, something like this:

`searchParams.toString()  = 'page=1&brand=nike&brand=adidas&orderBy=date'`

In this case, yo could do something similar with the query parameter or the router in Next to get this string, but it doesn't have a toString() method as URLSearchParams has, so you'd have to create it yourself
```typescript
export function parsedUrlQueryToURLString(query: ParsedUrlQuery): string {
  const params = []
  for (const [key, value] of Object.entries(query)) {
    if (!value) continue
    if (Array.isArray(value)) {
      value.forEach((element) => {
        params.push(`${key}=${encodeURIComponent(element)}`)
      })
    } else {
      params.push(`${key}=${encodeURIComponent(value)}`)
    }
  }
  return params.join('&')
}
```
I don't know how the rest of the people are doing, but I just use the ParsedUrlQuery with some helper functions that I made myself, at least until the guys in Vercel decide to switch to URLSearchParams or anything else.

**Reply by icyJoseph** [reply]
Using the snippet I posted I see this in `requestbin`, for this input, `/?foo&bar=joe&hello=world,all&bar=jane,name`

![Screenshot 2022-06-28 at 15 26 31](https://user-images.githubusercontent.com/21013447/176190350-29f1b18c-bba6-4699-9a2e-4be3f1969b67.png)

I think one thing to remember is that how multiple values are to be defined, is not standard, so your application, kind of has to agree on something, as you see, `bar=jane,name` is one value, and `bar=joe` is another, that's why the list is length 2, with `joe`, and `jane,name`.

**Reply by icyJoseph** [reply]
Also a JSON payload might not be so bad, though it requires perhaps a bit more than you'd like.

**Reply by florianwalther-private** [reply]
Thank you. For now, I'm going with the `parsedUrlQueryToURLString` you provided 👍. Although a JSON payload is something I will think about.
Regarding multiple values. Both `URLSearchParams` as well as Express' `SearchQueryParams` handle multiple values not as a comma-separated list, but as the same key appearing multiple times. Like `?languages=en&languages=fr`. So that's what I went with.

**Reply by ramirezsandin** [reply]
Well, I've checked in the [Node documentation about querystring ](https://nodejs.org/api/querystring.html#querystringstringifyobj-sep-eq-options)and there's a method called stringify that does exactly that (no need for custom helper functions), so better if you do:

```typescript
import { stringify } from 'querystring'

const parsedURLQueryParams = stringify(query)
```
viel Spaß!

**Reply by florianwalther-private** [reply]
Amazing, thank you very much! I wonder why that's not a method on `ParsedUrlQuery`. Would make it easier to find.

**Comment by ramirezsandin** [comment, marked as answer]
Long story short:

**Convert ParsedUrlQueryParams to encoded URL string**
Use the methods provided in[ querystring library of Node](https://nodejs.org/api/querystring.html#querystringstringifyobj-sep-eq-options) (Legacy but currently used in Next).
```typescript
import { encode } from 'querystring'

// From router with useRouter hook:
// const { query } = useRouter()

// From context in getServerSideProps_
// const { query } = ctx

const parsedURLQueryParams = encode(query)
```

**Convert From ParsedUrlQuery  to URLSearchParams**
Use the previous method and pass it to the URLSearchParams constructor
```typescript
import { encode } from 'querystring'

// From router with useRouter hook:
// const { query } = useRouter()

// From context in getServerSideProps_
// const { query } = ctx

const searchParams = new URLSearchParams(encode(query))
```

**Reply by GabenGar** [reply]
Be aware that `query` from router includes the keys from the dynamic paths segments, which will result in a different URL depending on how you pass it.
